R/xml_log.R

```diff
@@ -38,65 +15,93 @@ deepstate_xmlog <- function(log){
+    msg <-if(length(xwhat)){
+      xml_text(xml_find_all(err.node,".//xwhat//text"))
     }else{
-      msg <- get_string(err.node,"what")  
+      get_string(err.node,"what")  
     }
     stack.nodes <- xml2::xml_find_all(err.node, ".//stack")
-    #print(length(stack.nodes))
-    #print(stack.nodes[2])
     add.ret="No Address Trace found"
     address <- xml_text(xml_find_first(err.node,".//auxwhat"))
     if(!is.na(address)){
       rs <- xml_child(err.node,".//auxwhat")
       add.ret<-address.trace(err.node)
-      #print(add.ret)
-      if(!is.null(add.ret) && add.ret == 1){
-        print("entered")
+      if(!is.null(add.ret) && add.ret == "stack"){
         add.ret = stack.trace(stack.nodes[length(stack.nodes)],src.dir.xpath)
       }else{add.ret="No Address Trace found"}
     }else{address="No Address found"}
     stack.nodes <- xml2::xml_find_all(err.node, ".//stack")
-    count = 0
-    ret="NA"
     for(s.node in stack.nodes){
       ret = stack.trace(s.node,src.dir.xpath)
-      if(length(ret)){
-        count = count + 1
+      if(!is.na(ret)){
+        line.num.dt.list[[length(line.num.dt.list)+1]] <- data.table(
+          err.kind=kind,
+          message=msg,
+          file.line=ret,
+          address.msg=address,
+          address.trace=add.ret)  
       }
-      #if(gsub(" ","",add.ret) == ":"){address.trace="No address trace found"}
-      #print(add.ret)
-      line.num.dt.list[[length(line.num.dt.list)+1]] <- data.table(
-        err.kind=kind,
-        message=msg,
-        file.line=ret,
-        address.msg=address,
-        address.trace=add.ret)
-      if(count == 1){break}
     }
   }
-  #print(line.num.dt.list)
-  (line.num.dt <- do.call(rbind, line.num.dt.list))
+  line.num.dt <- do.call(rbind, line.num.dt.list)
   line.num.dt<-unique(line.num.dt, incomparables = FALSE)
-  return(line.num.dt)
-  
+}
+
+##' @title identify the stack trace with error
+##' @param s.node stack frame
+##' @param src.dir.xpath gives path to the src file
+##' @return trace with the error
+##' @export
+stack.trace <- function(s.node,src.dir.xpath){
+  first.dir <- xml2::xml_find_first(s.node, src.dir.xpath)
+  first.frame <- xml2::xml_parent(first.dir)
+  first.file <- get_string(first.frame, "file")
+  first.line <- get_string(first.frame, "line")
+  trace <- if(length(first.line) > 0 && length(first.file) >0){
+    line.trace <- paste0(first.file, ' : ', first.line)
+  }else{
+    line.trace <- NA_character_
+  }
+}
+
+##' @title identify the address trace with error
+##' @param err.node error node
+##' @return stack if stack trace found else the aux  
+##' @export
+address.trace<-function(err.node){
+  childs <- as_list(err.node)
+  names.list <- c(names(childs))
+  if(any(names.list=="auxwhat")){
+    vals <- which(names.list=="auxwhat")
+    val = (names.list[vals+1] == "stack")
+    output <- if(!is.na(val) && val == TRUE)
+      val <- "stack"
+  }else{
+    val <- "aux"
```

Comment:
please remove `val <-` since the value will just be returned.
also if this if statement fails the return value will be NULL, is that what you intended?

Reply:
Yes, aux means there is no stack trace present. Updated the else condition to return NA_character_.